### ISSUE TITLE
'venv --clear' should prompt user before nuking entire directory

### ISSUE THREAD

**Issue by b9fff05e-d562-4fe1-9b8f-e041086e1dc5** [python/cpython#90484]
BPO | [46326](https://bugs.python.org/issue46326)
--- | :---
Nosy | @vsajip, @eryksun, @alimpfard

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-01-10.09:01:46.744>
labels = ['type-bug']
title = "'venv --clear' should prompt user before nuking entire directory"
updated_at = <Date 2022-01-13.15:29:41.064>
user = 'https://github.com/alimpfard'
```

bugs.python.org fields:
```python
activity = <Date 2022-01-13.15:29:41.064>
actor = 'ned.deily'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2022-01-10.09:01:46.744>
creator = 'alimpfard'
dependencies = []
files = []
hgrepos = []
issue_num = 46326
keywords = []
message_count = 6.0
messages = ['410192', '410223', '410228', '410229', '410230', '410282']
nosy_count = 3.0
nosy_names = ['vinay.sajip', 'eryksun', 'alimpfard']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue46326'
versions = []
```

</p></details>

**Comment by vsajip** [comment]
I think venvs should be seen as disposable, and easily recreatable using e.g. `pip install -r requirements.txt`. What's the reason why this can't be done on the odd occasion when a venv is cleared out unintentionally? The help for this option is pretty unambiguous.

**Comment by crides** [comment]
was messing with `venv`, and now my (fortunately termux) `$HOME` is gone.

The help for the `venv` cli says:

```
positional arguments:
  ENV_DIR               A directory to create the environment in.
```

which means it should be perfectly fine to pass `.` or any existing directory as `ENV_DIR`, and same for `--clear` as well. You can say it's the user's (my) fault of not understanding what `venv` does, but also seems the help also doesn't say what it does clearly.

**Comment by c0m4r** [comment]
That's one of the evilest things I can imagine.

```
  --clear               Delete the contents of the environment directory 
                        if it already exists, before environment creation.
```

Does the description of this damn option says that it will wipe my source code? **YES**

Is it my fault that I just deleted an hour of my work? **certainly**

This can, of course, be understood literally, but for the love of god, please make it more **clear** and intuitive, because it may well be interpreted as clearing contents previously created by the venv itself, and not everything, including user files.

So please, either add a data loss warning in the description or a user-acknowledged warning before executing the command as mentioned earlier.

**Comment by c0m4r** [comment]
Which part of me admitting that **it was totally my fault** you didn't understand?

I use it differently for convenience and because I can. It's not like someone will come to me and slap me in the face because whooo I don't follow instructions. At most, I will wipe some part of my work, as I did, but I will forget about it 5 minutes later since I'm doing my best to backup or push any changes frequently; no big deal. Why is it convenient for me you ask? Because switching back and forth to the venv every single time is annoying, and I do it a lot, often on a fresh OS where I can't simply script it. Therefore, typing f.e. bin/python is quicker than .venv/bin/python. Does typing a few more characters seem like a hassle? No. Can I just switch to the venv and type even fewer characters? Yes. But it's convenient for most use cases, not every single one.

Not a good practice? On every single electronic device, there's a freaking warning sign and a message telling me that I'm going to die the moment I open it up, and it's not considered to be a good practice to do so. Yet, I'm still alive. There's a reason why this issue exists in the first place. It is because people sometimes don't follow instructions, and I'm taught to take that into account when designing anything, making stuff failsafe or at least prompting people for confirmation when I'm about to ruin their day, even though they didn't follow instructions. Like these warnings put on the electronics. That's why I am only suggesting, like the OP that it's not at least an intuitive and sufficient way to inform someone that you're about to wipe their data, and the option itself is counter-intuitive. Not mentioning that part:

> Simply deleting an entire directory tree with a command that's extremely prone to misunderstanding should not be virtualenv's job

Over and out

**Comment by vsajip** [comment]
> Which part of me admitting that it was totally my fault you didn't understand?

The part where you're complaining here.

**Comment by tusharsadhwani** [comment]
> Therefore, typing f.e. bin/python is quicker than .venv/bin/python

Consider investing in an auto venv activator like [Yelp/aactivator](https://github.com/Yelp/aactivator), and making your venvs disposable again.

**Comment by alimpfard** [comment]
Bit of a late comment, but allow me to clarify some things and respond to some comments.

> What's the reason why this can't be done on the odd occasion when a venv is cleared out unintentionally?

This is not at all about invoking `venv --clear` with the intent of refreshing and/or recreating a virtual env, imagine this very real scenario (unfortunately not too uncommon with tools, see [1] for an example)

```console
# Invoked by {mistake, some tool, some _tool_, etc etc}
$ python -m venv ~
# `python -m venv --help` has clear, surely that just clears the env - PEBKAC of course
$ python -m venv --clear ~
# RIP.
```

Of course, none of this is directly venv's fault (except maybe the part where you can make a venv in a nonempty dir, but you can't get rid of it) - the cli help is...minimally...clear on what the option does.

> [not] good practice to put your source code in a venv directory.

Certainly, and I have no problem with that, I imagine anyone doing so _intentionally_ is doing it at their own discretion.


Of course, I understand that the number of people facing this issue is not _nearly_ large enough to warrant introducing a breaking change (even if I personally don't agree with this option' behaviour), so perhaps it's better to add an option to actually purge venv files, and add that to the cli. win-win.

[1]: https://youtrack.jetbrains.com/issue/DS-3370/Welcome-pages-default-venv-directory-should-not-be-the-users-home-directory

**Comment by tusharsadhwani** [comment]
I actually don't mind a flag like, say `--only`, which only deletes the `bin`, `lib` and `pyvenv.cfg` files, and then only deletes the entire folder if it is empty.

Ideally that should have been the behaviour in `--clear` since the beginning, so not very sure if adding an opt-in flag for safety really help anyone though.

**Comment by alimpfard** [comment]
> adding an opt-in flag for safety

Yes, adding an opt-in complementary flag is unlikely to help anyone.
I can't say for sure, but I think someone only skimming the help output/page is more likely to notice two flags (e.g. `--clear` and `--purge`) for seemingly the same purpose, and thus *actually* pay attention to the description of those flags.

**Comment by metamaterialsuit** [comment]
Totally agree that the behavior of venv --clear clearly does not follow the principle of least surprise. Furthermore, it does zero validation or sanity checking of its arguments before performing a destructive and irreversible operation on a users filesystem. 
To add insult to injury, determining whether or not a virtual environment has actually been specified is trivial. venv creates a pyenv.cfg file in the virtual environment, which means just a basic check for that file would make the command safe.

**Comment by tusharsadhwani** [comment]
> venv creates a pyenv.cfg file in the virtual environment, which means just a basic check for that file would make the command safe.

Not safe if you accidentally create a venv in your `$HOME` directory as shown in an example above, or if you like to have your source files inside the venv for some reason, which I've also seen in the wild.

**Comment by metamaterialsuit** [comment]
You're right. I think clear needs some combination of an additional confirmation from the user, far better documentation, etc, in addition to at least SOME effort to verify it's doing the correct thing. Also, I wonder, does venv always have the same structure? I.e. some folders for packages and the like + some venv generated metadata? If venv knows everything that will be in the venv, then venv should simply delete that!
I truly cannot believe that this tool will simply delete an entire file tree. This tool is designed to manage 
**virtual environments**, thats it.

**Comment by hugovk** [comment]
I see uv and virtualenv do the same thing, although perhaps a key difference is uv defaults the path to `.venv`:

* https://docs.astral.sh/uv/pip/environments/
* https://docs.astral.sh/uv/reference/cli/#uv-venv

Perhaps venv should also default to `.venv` to help mitigate this, so the user doesn't have to pick a directory that may cause problems later.

**Comment by alimpfard** [comment]
> Perhaps venv should also default to .venv

I'm not sure how that helps when a path is explicitly passed in a la https://github.com/python/cpython/issues/90484#issuecomment-2187680899?

the reality of the matter is that `--clear` is just not what people expect, and people are generally very bad at seeing things they don't expect to see in help outputs.

**Comment by hugovk** [comment]
It would help in that, right now, you get an error when running `python3 -m venv` without specifying a directory:

```
python3 -m venv: error: the following arguments are required: ENV_DIR
```

Which forces you to pick something, and you might choose something that causes problems later when you run `--clear` without knowing what it does.

If we instead default to `.venv`, you can more safely run `--clear`.

uv defaults to `.venv` and I couldn't find any reports on their tracker or in their Discord that people accidentally deleted a poorly chosen dir.

**Comment by alimpfard** [comment]
:shrug: it just feels like you're trying to "fix" the problem by inventing a way to hide it. also:
> unfortunately not too uncommon with tools, see [1](https://youtrack.jetbrains.com/issue/DS-3370/Welcome-pages-default-venv-directory-should-not-be-the-users-home-directory) for an example

I suppose anything is better than the nothing we've had so far.

**Comment by metamaterialsuit** [comment]
Have we made any progress here? Given the number of people who have complained thus far it seems likely somebody else is going to be in for a really bad day if we don't do anything.
